client/lib/perfmon/index.js

```diff
@@ -0,0 +1,173 @@
+/** 
+ * Internal dependencies
+ */
+import analytics from 'analytics';
+
+/**
+ * External dependencies
+ */
+import each from 'lodash/collection/each';
+import remove from 'lodash/array/remove';
+import debounce from 'lodash/function/debounce';
+import page from 'page';
+
+var debug = require( 'debug' )( 'calypso:perfmon' );
+
+const PLACEHOLDER_CLASSES = [
+	'placeholder',
+	'pulsing-dot is-active',
+	'is-loading'
+];
+
+const PLACEHOLDER_MATCHER = PLACEHOLDER_CLASSES.map(function(clazz) { return `[class*='${clazz}']`; }).join(', ');
+const OBSERVE_ROOT = document.getElementById('wpcom');
+
+let activePlaceholders = [];
+let placeholdersVisibleStart = null;
+let initialized = false;
+
+// add listeners for various DOM events - scrolling, mutation and navigation
+// and use these to trigger checks for visible placeholders (and, in the case of mutations,
+// to record new placeholders and remove nodes that are no longer placeholders)
+function observeDomChanges( MutationObserver ) {
+	// if anything scrolls, check if any of our placeholder elements are in view,
+	// but not more than a few times a second
+	window.addEventListener('scroll', debounce(checkForVisiblePlaceholders.bind(null, 'scroll'), 200), true);
+
+	// if the user navigates, stop the current event and proceed to the next one
+	page( function( context, next ) {
+		// send "placeholder-wait-navigated" event
+		checkForVisiblePlaceholders( 'navigate' );
+		next();
+	} );
+
+	// this is fired for matching mutations (childList and class attr changes)
+	var observer = new MutationObserver(function(mutations, observer) {
```

Comment:
We should probably either use `let` in these places or move the `var` to the top of this function according to our style guide. I'm pretty sure we do it that way because the variables are hoisted up there anyway internally.